File: 39-combination-sum/39-combination-sum.scala
import scala.collection.mutable._

object Solution {
  def combinationSum(candidates: Array[Int], target: Int): List[List[Int]] = {
    val sorted = candidates.sorted
    helper(target, sorted, Map())
  }

  def helper(
      num: Int,
      cand: Array[Int],
      map: Map[Int, List[List[Int]]],
  ): List[List[Int]] = {
    val res = map.get(num).getOrElse {
      var res: List[List[Int]] = List()
        
      cand.foreach { c =>
        if (num == c) { res = res :+ List(c) } 
        else if (num > c) {
          val sub = helper(num - c, cand, map)
          sub.collect {  case s if s(0) >= c => c +: s }.foreach { s => res = res :+ s }
        }
      }
        
      res
    }
      
    map += (num -> res)
    res
  }
}

File: 39-combination-sum/39-combination-sum.java
class Solution {
    List<List<Integer>> res = new ArrayList<>();
    
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        helper(target, candidates, 0, new LinkedList<>());
        return res;
    }
    
    void helper(int target, int[] candidates, int start, LinkedList<Integer> current) {
        if (target == 0) {
            res.add(List.copyOf(current));
            return;
        }
        if (target < 0) return;
        if (start >= candidates.length) return;
        
        for(int i = start; i < candidates.length; i++) {
            current.add(candidates[i]);
            helper(target - candidates[i], candidates, i, current);
            current.removeLast();
        }
    }
}